### DISCUSSION TITLE
Setting quay.expires-after via API not allowed

### DISCUSSION THREAD

**Discussion by enyineer** [quay/quay#1405] [Q&A] [answered]
Hi all,

I do have the following problem: I need to set the expiry tag via API for a specific manifest using the POST /api/v1/repository/{repository}/manifest/{manifestref}/labels endpoint. This is because our automatic Housekeeping concept was written so that the further an Image gets promoted forward in stages, the longer it should be kept in our Quay Enterprise (On-Prem) instance before being garbage collected. Eg: Snapshots build for testing should not be kept as long as builds going to production.

The problem here is: Every build can potentially go to production, so setting this Label via Dockerfile is not an option. It needs to be adjusted via API according to the stage it got promoted to.

Setting labels via the above endpoint works fine, except for protected key prefixes like "quay".expires-after (as configured in [config.py](https://github.com/quay/quay/blob/56b16b70cca1ff0f5f51dd2fb47852027d66eef6/config.py#L93)).

Is there any way to disable this protection? It does not make sense in my opinion to not allow setting this tag via API.

Thanks in advance!

Cheers
Nico

**Comment by dmesser** [comment, marked as answer]
@enyineer why not simply use the Quay API to update the expiration like so: https://access.redhat.com/documentation/en-us/red_hat_quay/3.7/html/red_hat_quay_api_guide/red_hat_quay_application_programming_interface_api#changetag

**Reply by enyineer** [reply]
Thanks for the ultra fast reply. I'll give that a try and will get back to you if any problems arise. Wouldn't it make sense though to be able to edit that tag anyways via the labels endpoint? I do understand the need of protecting specific prefixes but using the changeTag method requires more information than just the digest. I would be happy about a possibility to whitelist specific labels that should be allowed to be changed via API even though they're protected by default.

**Reply by enyineer** [reply]
I take back the "requires more information" part because needing both digest and tag name seems to be optional.

**Reply by enyineer** [reply]
But one thing is still inconvenient: If I do have multiple tags pointing to the same digest, I would have to do this for all tags that exist instead of defining the expiration for the image itself and thus all tags connected to it.

**Reply by dmesser** [reply]
I guess we want to still have the ability to do some validation on the data vs. arbitrarily accepting any input in labels with special meanings. E.g. a negative value for expiration date or something in the past. Hence the way to do this via the proper Quay API is preferred rather than directly editing the data storage.

**Reply by enyineer** [reply]
I do understand. Thanks for pointing this out. Have a nice day!

**Reply by enyineer** [reply]
@dmesser one last question. What would be a valid "expiration" value? The documentation does not seem to show examples and "90w" is not a valid input as it only accepts numbers (in contrast to the expires-after label). Is it hours, days, weeks, ...?

**Reply by dmesser** [reply]
That's a gap in our docs. AFAIK it's a UNIX time stamp of the expiry date. In such cases I always do the desired action via the Quay UI and observe what backend is called and how the values are provided.

**Reply by enyineer** [reply]
Unix Timestamp works perfectly fine.

**Reply by enyineer** [reply]
@dmesser I just realized there's probably a bug when setting an expiration date via UI for all tags in a repository. If you set a specific date eg. 22nd of June 2024 (in two years) tags that have not been pushed today will have the expiration date set before 22nd of June 2024.
If there is, for example, a tag in the list which was built to years ago, it is immediately scheduled for garbage collection. Is that wanted? It at least seems to not match the input fields description which says "If specified, the date and time that the key expires. If set to none, the tag(s) will not expire.". This would leave me to believe that all tags will expire at that date, not the timespan from current date to the set date minus the timespan from push to current date.

**Reply by flavianmissi** [reply]
The behaviour you're describing is indeed not how it should work. Could you provide some steps on how to reproduce? I'm getting the intended behaviour when simply setting expiration in bulk via UI.
